server/main-api/src/search/search_executor/lexer.rs

```diff
@@ -0,0 +1,167 @@
+    let slice = lex.slice();
+    let mut split = slice.len();
+    for (i, c) in slice.char_indices().rev() {
+        if c.is_ascii_digit() {
+            split = i;
+        } else {
+            break;
+        }
+    }
+    let (text, numbers) = slice.split_at(split);
+    (text.to_string(), numbers.to_string())
+}
+
+/// Removes the specified prefix and additional whitespace from the token
+/// e.g used to remove the "in:" and "@" prefixes from filters
+fn remove_prefix(lex: &mut Lexer<Token>, prefix: &'static str) -> String {
+    lex.slice()[prefix.len()..].trim().to_string()
+}
+
+/// Parses the query string into a list of tokens
+/// priority between tokens is set as follows
+/// Filters|QuotedText > SplittableText > Text > skip
+#[derive(Logos, Debug, PartialEq, Eq, Clone)]
+#[logos(skip r"[ \t\n\f]+")]
+pub enum Token {
+    #[regex("\"[^\"]+\"", | lex | lex.slice()[1..lex.slice().len() - 1].to_string(), priority = 3)]
+    #[regex(r"[^ \t\n\f]+", | lex | lex.slice().to_string())]
+    Text(String),
+
+    #[regex("[a-zA-Z]+[0-9]{1,4}", irregular_split, priority = 2)]
+    SplittableText((String, String)),
+
+    #[regex("in: ?[a-zA-Z0-9]+", | lex | remove_prefix(lex, "in:"), priority = 3)]
+    #[regex("@ ?[a-zA-Z0-9]+", | lex | remove_prefix(lex, "@"), priority = 3)]
+    ParentFilter(String),
+
+    #[regex("usage: ?[a-zA-Z0-9]+", | lex | remove_prefix(lex, "usage:"), priority = 3)]
+    #[regex("nutzung: ?[a-zA-Z0-9]+", | lex | remove_prefix(lex, "nutzung:"), priority = 3)]
+    #[regex("= ?[a-zA-Z0-9]+", | lex | remove_prefix(lex, "="), priority = 3)]
+    UsageFilter(String),
+
+    #[regex("type: ?[a-zA-Z0-9]+", | lex | remove_prefix(lex, "type:"), priority = 3)]
+    TypeFilter(String),
+}
+
+#[cfg(test)]
+mod tokenizer_tests {
+    use super::*;
+
+    #[test]
+    fn empty() {
+        assert_eq!(Token::lexer("").next(), None);
+        assert_eq!(Token::lexer("\t").next(), None);
+        assert_eq!(Token::lexer("\n").next(), None);
+        assert_eq!(Token::lexer("  ").next(), None);
+    }
+
+    #[test]
+    fn quoting() {
+        let mut lexer = Token::lexer("\"");
+        assert_eq!(lexer.next(), Some(Ok(Token::Text("\"".to_string()))));
+        assert_eq!(lexer.next(), None);
+
+        let mut lexer = Token::lexer("\"\"");
+        assert_eq!(lexer.next(), Some(Ok(Token::Text("\"\"".to_string()))));
+        assert_eq!(lexer.next(), None);
+
+        let mut lexer = Token::lexer("\" \"\"");
+        assert_eq!(lexer.next(), Some(Ok(Token::Text(" ".to_string()))));
+        assert_eq!(lexer.next(), Some(Ok(Token::Text("\"".to_string()))));
+        assert_eq!(lexer.next(), None);
+        for text in ["a", "a ", "a a ", " a a ", " @ = in: contains: type: a "] {
+            let quoted_text = format!("\"{text}\"");
+            let mut lexer = Token::lexer(&quoted_text);
+            assert_eq!(lexer.next(), Some(Ok(Token::Text(text.to_string()))));
+            assert_eq!(lexer.next(), None);
+        }
```

Comment:
Do you know what happens for the special cases `in:"`, `@"Garching"`, `usage: "lecture hall"` etc? I think allowing quoting for filter values would be a valuable feature, but currently they have the same priority
(I can test this weekend maybe)

Reply:
Reason for the priority being the the same is, that `"..."` and `in:...` have a different start => can never intersect

I think this is a special case which we currently probably do not need.
I have added a testcase documenting this behaviour.

Examples:
`in:foo` -> `ParentFilter("foo")`
`in: foo` -> `ParentFilter("foo")`
`in:"foo"` -> `Text("in:"foo"")`
`in: "foo` -> `Text("in:")`, `Text(""foo")`
`in: "foo"` -> `Text("in:")`, `Text("foo")`

Reply:
Proposed Solutions:
- if a usage includes spaces, we should encode it differently (e.g. `lecture`_hall`)
- given that the others are also only direct-match filters, this should not be nessesary

Reply:
I am merging this PR, as the problems noticed here are not major enough imo